### DISCUSSION TITLE
Better Moderation Tools

### DISCUSSION THREAD

**Discussion by hubismal** [FakeFishGames/Barotrauma#4675] [Ideas and suggestions]
Specifically making this just to remind devs to **implement post-disconnect banning/user actions**, but people are welcome to suggest more ideas to make this kind of a megaissue for moderation(?).

**Comment by Regalis11** [comment]
This is currently being worked on (ticket in the private repo: https://github.com/Regalis11/Barotrauma-development/issues/1709). Specifically, you can ban clients after they've disconnected by selecting them from the crew list, server log or the chatbox.

**Comment by Jlobblet** [comment]
**Import Banlist Button**

Right now the only way to add people to your banlist is to either ban them ingame when they show up, or to manually add them. The game could have a button to add all people on a selected banlist to yours, making it easier for server moderators who want to share their banlist with others.

**Comment by Cogs16** [comment]
A language filter for kicking those who use racial slurs automatically would be a great help.

**Reply by MystLeissa** [reply]
I'd disable it....Stick and Stones can break my bones, but words only hurt if not spoken.

**Reply by MystLeissa** [reply]
There's nothing like being forced to shut up when you have something to say because someone else views it as "Morally Unacceptable". It can drive people to commit far greater crimes; and I would know, I've been the one silenced, and have lashed out in the past because of it. I don't like it, but it's human nature.

**Reply by Cogs16** [reply]
I understand how not being able to voice a legitimate complaint would be annoying. However as a host, I do not see how a filter designed to prevent one-word racial/homophobic slurs(and maybe swearing), would impair a legitimate complaint.  I don't see why non-constructive criticism should be really tolerated. At the end of the day though, such a system would be optional for server owners.

**Comment by MystLeissa** [comment]
While we're talking about Ban Lists, could you do something about (legitimate) players being forced into situations where when they d/c they die before they can reconnect? I feel like if one is persistent the other should be as well.

**Comment by Krzeszny** [comment]
The server log needs a Trusted filter to prevent it from showing actions from trusted players.

**Comment by trevormoney1** [comment]
Add a permission, to be given to moderators by default, to be able to see through disguises whilst spectating. That would be so nice.

**Reply by ghost** [reply]
logs already see through disguises, no?

**Comment by ghost** [comment]
To be honest, allowing the host to deny server access for Steam accounts with
- VAC bans (any)
- Steam profile bound game bans (any)
- private/bugged/uninitialized Steam profiles
is expected to solve most spectacular problems with as little effort as possible.
And it might be just enough. In the end, if the host is picky about players - the host can assemble a party for private server in advance.

(I'd really like an option to hide servers with such "players" present from servers list too)

Upd. Found my old document (I planned to TRY to address this problem with mod or something - but something lured me away from this idea). I hope that this exceprt would provide some interesting insights. Do note that I was focused on preventing mishaps, not on post-moderation.

High risk player group (for Serious Roleplay): if any is true
- reached maximum negative Karma at any given time during server's existance
- Steam profile does not exist or does not return valid intel on following API calls
- have any VAC ban
- have any game ban observe'able through Steam profile
- have active mute in Barotrauma's Steam community
- account created less than week ago, owns only Barotrauma (and any amount of Free games - these do not contribute towards owned games count, exploitable for our purposes)
- account created more than month ago, owns only Barotrauma (and any amount of Free games), does not have any Tutorial achievements ("As Good As New" should suffice) in it

Med risk player group (for Casual): if any is true
- reached 50% negative Karma at any given time during server's existance
- Steam profile does not exist or does not return valid intel on following API calls
- have any VAC ban
- account created more than month ago, owns only Barotrauma (and any amount of Free games), does not have any Tutorial achievements ("As Good As New" should suffice) in it
- account created more than year ago, owns only Barotrauma (and any amount of Free games), Steam account level is less than 6

**Reply by JoneKone** [reply]
I suggest you do this mod in Lua with Evil's Lua for Barotrauma  I'm pretty sure all of this is very achievable on the lua platform. all server side.

If you ever make it, please share with me :)

**Reply by ghost** [reply]
Sorry for a late response. I've left GitHub some time ago. Returned just to post few issues for stuff which finally made me explode mentally. And FINALLY seen your response #thanxgithubforcrappyui

Since LUA support is being shipped as executable replacer, it's a big no-no for me. I don't believe in end users honesty and fair use, you know, and executable replaceables are a reliable attack method. Also, compatibility AND necessity to rebuild the app on each game update.

If (if!) I'd ever gonna try to do such an app, I'll probably resort to something more robust.
Like parsing server logs for connecting user's SteamID (for server) or fellow players' SteamIDs (for client, upon joining, if possible at all).
Then checking if he's/they're in any 'risk' group.
Then auto-banning him (for server) or auto-disconnecting (for client).
All as a completely stand-alone application which is not really dependant on Barotrauma executable (if possible).

As LUA needs learning some coding anyway, the only real disadvantage of this new scheme is necessity to keep different set of tricks for different OS families (for Linux: also for different variants).
I'm not really very savvy with both Steam API and client-server connection relationships, however, this scheme looks simple enough to implement.

This idea is free for taking. No credit needed - the good you'll do with its implementation is enough for me.
